### PR TITLE
Add blinking cursor as a "responding" indicator

### LINKED ISSUE #55184: JetBrains: Chat: Indicate "responding" state
We're currently not clearly indicating what's happening after the user sent a message or invoked a recipe.

Data points on how others solved this:
- In VS Code, we have blinking cursor displayed before we get any response from Cody
- Same in ChatGPT

So a blinking filled rectangle might be a good idea here.

### PR DESCRIPTION
**Pull request by mkondratek**
This PR adds a "responding" indicator to the chat. 

The indicator is shown after sending a message and before the first words of the response.

Fixes https://github.com/sourcegraph/sourcegraph/issues/55184.

## Demo

https://github.com/sourcegraph/jetbrains/assets/19799111/6803b03c-86ae-4243-9e04-0e09b529d88f

## Test Plan
- [x] 1. Validation of the expected process
   1. Launch the `runIde` command.
   2. Confirm that the initial hello message is displayed without any indicator.
   3. Input a question and select `Send`.
   4. Ensure that the question appears in the chat window.
   5. Verify that the blinking cursor - indicating the response generation - is visible in the chat.
   6. Confirm that the indicator disappears once the response begins to display.
   7. Validate that the entire response is received and shown in the chat, without any indicator.

- [x] 2. Validation of the 'Stop generating' function
   1. Initiate the `runIde` command.
   2. Confirm that the hello message is visible, and no indicator is present.
   3. Input a question and select `Send`.
   4. Ensure that the question is added to the chat.
   5. Verify that the blinking cursor - indicating response generation - is visible in the chat.
   6. Test the `Stop generating` feature before the response begins to display.
   7. Confirm the disappearance of the indicator as expected.
   
- [x] 3. Validation of the themes
  1. Initiate the `runIde` command.
  2. Ensure that the indicator looks correct with the light theme.
  3. Ensure that the indicator looks correct with the dark theme.

### CODE REVIEW

File: src/main/java/com/sourcegraph/cody/chat/BlinkingCursorComponent.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  private BlinkingCursorComponent() {
+    this.showCursor = true;
+    this.timer =
+        new Timer(
+            500,
+            e -> {
+              showCursor = !showCursor;
+              repaint();
+            });
+    this.timer.start();
```
Comment (mkondratek) [inline src/main/java/com/sourcegraph/cody/chat/BlinkingCursorComponent.java]: The tests fail. In CI the info is very brief. In a local run (`./gradlew test`) I see the following error:

```
junit.framework.AssertionFailedError: Not disposed javax.swing.Timer: Timer (listeners: [com.sourcegraph.cody.chat.BlinkingCursorComponent$$Lambda$908/0x0000000800a00840@160f9d46]) (delayed for 379ms); queue:TimerQueue (javax.swing.Timer@970746c)
	at com.intellij.testFramework.TestApplicationManagerKt.checkJavaSwingTimersAreDisposed(TestApplicationManager.kt:355)
	at com.intellij.testFramework.TestApplicationManagerKt.tearDownProjectAndApp(TestApplicationManager.kt:207)
	at com.intellij.testFramework.fixtures.impl.LightIdeaTestFixtureImpl.lambda$tearDown$3(LightIdeaTestFixtureImpl.java:79)
	at com.intellij.testFramework.RunAllKt.doRun(runAll.kt:50)
	at com.intellij.testFramework.RunAllKt.access$doRun(runAll.kt:1)
	at com.intellij.testFramework.RunAll.run(runAll.kt:17)
	at com.intellij.testFramework.fixtures.impl.LightIdeaTestFixtureImpl.tearDown(LightIdeaTestFixtureImpl.java:105)
	at com.intellij.testFramework.fixtures.impl.CodeInsightTestFixtureImpl.lambda$tearDown$33(CodeInsightTestFixtureImpl.java:1326)
	at com.intellij.testFramework.EdtTestUtil.runInEdtAndWait(EdtTestUtil.java:33)
	at com.intellij.testFramework.fixtures.impl.CodeInsightTestFixtureImpl.lambda$tearDown$34(CodeInsightTestFixtureImpl.java:1326)
	at com.intellij.testFramework.RunAllKt.doRun(runAll.kt:50)
	at com.intellij.testFramework.RunAllKt.doRun$default(runAll.kt:46)
	at com.intellij.testFramework.RunAll$Companion.runAll(runAll.kt:27)
	at com.intellij.testFramework.RunAll.runAll(runAll.kt)
	at com.intellij.testFramework.fixtures.impl.CodeInsightTestFixtureImpl.tearDown(CodeInsightTestFixtureImpl.java:1284)
	at com.intellij.testFramework.fixtures.BasePlatformTestCase.tearDown(BasePlatformTestCase.java:62)
	at com.intellij.testFramework.UsefulTestCase.invokeTearDown(UsefulTestCase.java:483)
	at com.intellij.testFramework.UsefulTestCase.defaultRunBare(UsefulTestCase.java:471)
	at com.intellij.testFramework.UsefulTestCase.lambda$runBare$12(UsefulTestCase.java:541)
	at com.intellij.testFramework.EdtTestUtil.lambda$runInEdtAndWait$1(EdtTestUtil.java:40)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:303)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:776)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:746)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:393)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
```

**Comment by danielmarquespt** [comment]
UI wise looks good! Nice improvement

File: src/main/java/com/sourcegraph/cody/chat/BlinkingCursorComponent.java (hunk 1 above)
Comment (olafurpg) [reply src/main/java/com/sourcegraph/cody/chat/BlinkingCursorComponent.java]: Before seeing your comment, I was going to check if we dispose this timer. I'm surprised our CI caught this, I thought we don't have enough test coverage to catch these kinds of bugs.

**Review by olafurpg — APPROVED** [review]
I haven't tried this locally but the video looks great 🤩

**Review by exigow — APPROVED** [review]
Looks good :+1: 

I've played with it locally. This is definitely improvement from UX perspective. It also behaves identically to VSCode.

**Comment by exigow** [comment]
You might need to rebase it to latest main due to failing test (not related).

**Review by Sa1to — APPROVED** [review]

File: src/main/java/com/sourcegraph/cody/CodyToolWindowContent.java [hunk 2]
```diff
@@ -522,7 +538,7 @@ private void sendMessage(
 
     String displayText = XmlStringUtil.escapeString(message);
     ChatMessage humanMessage = new ChatMessage(Speaker.HUMAN, message, displayText);
-    addMessageToChat(humanMessage);
+    addMessageToChat(humanMessage, true);
```
Comment (pkukielka) [inline src/main/java/com/sourcegraph/cody/CodyToolWindowContent.java]: I don't know Sourcegrph code standards, but maybe when using boolean value from java we could add a variable name?

`addMessageToChat(humanMessage, /* shouldDisplayBlinkingCursor = */ true)`

In the long term just `true` or `false` is hard to read and reason about.

Comment (mkondratek) [reply src/main/java/com/sourcegraph/cody/CodyToolWindowContent.java]: good point, thanks

Comment (exigow) [reply src/main/java/com/sourcegraph/cody/CodyToolWindowContent.java]: > In the long term just true or false is hard to read and reason about.

+1

However, isn't this problem solved by inlay parameter names in IDE? Also, I'm not sure if the IDE will take such a comment into account during refactoring variable name.

IMO the most *clean* approach would be to split this into two methods (smth like addMessage and addBlinkingMessage).

Comment (mkondratek) [reply src/main/java/com/sourcegraph/cody/CodyToolWindowContent.java]: I moved a few classes to Kotlin - as we disscussed it in the other PR, it's much better approach :D

**Review by Sa1to — APPROVED** [review]

**Pull request merged**